packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx

```diff
@@ -90,6 +90,21 @@ describe('<DataGrid /> - Pagination', () => {
       expect(onPageChange.callCount).to.equal(2);
     });
 
+    it('should trigger onPageChange with correct page param when page prop is supplied i.e in controlled mode', () => {
+      const onPageChange = spy();
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} page={1} onPageChange={onPageChange} pageSize={1} />
+        </div>,
+      );
+
+      fireEvent.click(screen.getByRole('button', { name: /next page/i }));
+      expect(onPageChange.lastCall.args[0].page).to.equal(2);
+      fireEvent.click(screen.getByRole('button', { name: /previous page/i }));
+      expect(onPageChange.lastCall.args[0].page).to.equal(1);
```

Comment:
The state doesn't change, it should be:

```suggestion
      expect(onPageChange.lastCall.args[0].page).to.equal(0);
```

Reply:
> The state doesn't change

🤔  Umm, the state changes whenever the `onPageChange` is called. It changes here:
```js
setGridState((state) => ({
        ...state,
        pagination: applyConstraints({
          ...state.pagination,
          page,
        }),
      }));
```
I did not get you. If you click previous it will change from 2 to 1. How come directly `0`?

Reply:
> If you click previous it will change from 2 to 1. 

No, it doesn't change, the page stays on 1 because the state is controlled and forced on 1.

Reply:
So are you saying the `page` prop isn't for setting current page when the DataGrid is loaded initially? And when it is forced with the prop (hardcoded), the user can't paginate? And when `onPageChange` is applied, it still shouldn't paginate?

> the page always stays on 1 

And you suggested code change above from `1` to `0`, now you are saying page always stays on `1`... 

> The page doesn't change, the page always stays on 1 because the state is controlled and forced with the prop.

I am talking about `params.page`, are you saying that `params.page` doesn't change? I am testing parameters page.

Reply:
Your test is right in asserting that `params.page` is equal to 1. We need to ensure that it's called with the right value.

The concern raised by @oliviertassinari is valid, but I think we should address it in another PR, since it was already working this way before the regression. In controlled mode, there shouldn't be any change to the state. The `page` prop, and only it, controls the pagination. What we should do is to call `onPageChange` with the new page number, but is the developer who is responsible for updating the `page` prop with the value received. No `setGridState` calls in controlled mode.

Reply:
Okay. I was getting a bit confused by what Olivier meant. Also I apologise, if I sounded rude here in  https://github.com/mui-org/material-ui-x/pull/1719#discussion_r637062864. I am also trying my best to understand and fix the bug. 
I would say it is all fine now but only if we all agree.